backend/danswer/connectors/google_drive/connector_auth.py

```diff
@@ -0,0 +1,110 @@
+import os
+from typing import Any
+from urllib.parse import parse_qs
+from urllib.parse import urlparse
+
+from danswer.configs.app_configs import GOOGLE_DRIVE_CREDENTIAL_JSON
+from danswer.configs.app_configs import GOOGLE_DRIVE_TOKENS_JSON
+from danswer.configs.app_configs import WEB_DOMAIN
+from danswer.dynamic_configs import get_dynamic_config_store
+from danswer.utils.logging import setup_logger
+from google.auth.transport.requests import Request  # type: ignore
+from google.oauth2.credentials import Credentials  # type: ignore
+from google_auth_oauthlib.flow import InstalledAppFlow  # type: ignore
+from googleapiclient import discovery  # type: ignore
+
+logger = setup_logger()
+
+SCOPES = ["https://www.googleapis.com/auth/drive.readonly"]
+FRONTEND_GOOGLE_DRIVE_REDIRECT = f"{WEB_DOMAIN}/auth/connectors/google_drive/callback"
+google_drive_csrf_state: str = ""  # TODO save this to DB
+
+
+def backend_get_credentials() -> Credentials:
+    """This approach does not work for the one-box builds"""
+    creds = None
+    if os.path.exists(GOOGLE_DRIVE_TOKENS_JSON):
+        creds = Credentials.from_authorized_user_file(GOOGLE_DRIVE_TOKENS_JSON, SCOPES)
+
+    if not creds or not creds.valid:
+        if creds and creds.expired and creds.refresh_token:
+            creds.refresh(Request())
+        else:
+            flow = InstalledAppFlow.from_client_secrets_file(
+                GOOGLE_DRIVE_CREDENTIAL_JSON, SCOPES
+            )
+            creds = flow.run_local_server()
+
+        with open(GOOGLE_DRIVE_TOKENS_JSON, "w") as token_file:
+            token_file.write(creds.to_json())
+
+    return creds
+
+
+def get_drive_tokens(token_path: str = GOOGLE_DRIVE_TOKENS_JSON) -> Any:
+    if not os.path.exists(token_path):
+        return None
+
+    creds = Credentials.from_authorized_user_file(token_path, SCOPES)
+
+    if not creds:
+        return None
+    if creds.valid:
+        return creds
+
+    if creds.expired and creds.refresh_token:
+        try:
+            creds.refresh(Request())
+            if creds.valid:
+                with open(token_path, "w") as token_file:
+                    token_file.write(creds.to_json())
+                return creds
+        except Exception as e:
+            logger.info(f"Failed to refresh google drive access token due to: {e}")
```

Comment:
nit: should probably be logger.error or logger.exception. This might be useful in the future if we setup sentry for managed versions (or if people deploying this want to setup sentry)